internal/setup/setup.go

```diff
@@ -89,20 +92,29 @@ func Setup(ctx context.Context, config DBConfigProvider) (*serverenv.ServerEnv,
 		opts = append(opts, serverenv.WithBlobStorage(storage))
 	}
 
+	// Setup the database connection.
 	db, err := database.NewFromEnv(ctx, config.DB())
 	if err != nil {
 		return nil, nil, fmt.Errorf("unable to connect to database: %v", err)
 	}
 	opts = append(opts, serverenv.WithDatabase(db))
 
-	if apicfg, ok := config.(DBAPIConfigProvider); ok {
-		cfgProvider, err := dbapiconfig.NewConfigProvider(db, sm, apicfg.API())
+	// APIConfig must come after database setup due to the dependency.
+	if typ, ok := config.(APIConfigProvider); ok {
+		provider, err := apiconfig.NewDatabaseProvider(ctx, db, typ.APIConfigConfig(), apiconfig.WithSecretManager(sm))
 		if err != nil {
-			// APIConfig must come after DB due to dependency, ensure connection is closed
+			// Ensure the database is closed on an error.
 			defer db.Close(ctx)
 			return nil, nil, fmt.Errorf("unable to create APIConfig provider: %v", err)
 		}
-		opts = append(opts, serverenv.WithAPIConfigProvider(cfgProvider))
+
+		// Load initial configurations. This will help identify any
+		// misconfigurations.
+		if _, err := provider.AppConfig(ctx, ""); err != apiconfig.AppNotFound {
+			logger.Errorf("failed to load APIConfigs: %v", err)
```

Comment:
Open to suggestions. Should we return an error here? 

Pro: 

- Super obvious you have misconfiguration

Con:

- Bad config prevents new server starts (or scaling on serverless)

Reply:
I think we need to add a metric here so that this could be alerted on.

I’m coming down on the side of don’t return an error - this lets us control the return code that clients could get rather than getting an error from the infrastructure for container failed to start.

Reply:
Done (created metric, left as log)